openedx/core/djangoapps/external_auth/views.py

```diff
@@ -127,23 +129,31 @@ def _external_login_or_signup(request,
                               email,
                               fullname,
                               retfun=None):
-    """Generic external auth login or signup"""
+    """
+    Generic external auth login or signup
+    """
     # pylint: disable=too-many-statements
     # see if we have a map from this external_id to an edX username
-    try:
-        eamap = ExternalAuthMap.objects.get(external_id=external_id,
-                                            external_domain=external_domain)
-        log.debug(u'Found eamap=%s', eamap)
-    except ExternalAuthMap.DoesNotExist:
-        # go render form for creating edX user
-        eamap = ExternalAuthMap(external_id=external_id,
-                                external_domain=external_domain,
-                                external_credentials=json.dumps(credentials))
-        eamap.external_email = email
-        eamap.external_name = fullname
-        eamap.internal_password = generate_password()
+    eamap_defaults = {
+        'external_credentials': json.dumps(credentials),
+        'external_email': email,
+        'external_name': fullname,
+        'internal_password': generate_password()
+    }
+
+    # We are not guaranteed to be in a transaction here since some upstream views
+    # user non_atomic_requests
+    with transaction.atomic():
+        eamap, created = ExternalAuthMap.objects.get_or_create(
```

Comment:
Why not use outer_atomic in these cases too?  I'm not suggesting it, I just couldn't remember your rationale if you said it already.

Reply:
Just trying to narrow the scope of that fairly esoteric mode, but that doesn't really make sense since it will fail further down anyway. Better to fail up here and avoid the mismatched state, I'll change it and the one below for consistency. Good catch.